hpx/runtime/serialization/detail/polymorphic_centralized_factory.hpp

```diff
@@ -0,0 +1,218 @@
+#define HPX_SERIALIZATION_POLYMORPHIC_CENTRALIZED_FACTORY_HPP
+
+#include <hpx/runtime/serialization/serialization_fwd.hpp>
+#include <hpx/runtime/serialization/detail/polymorphic_intrusive_factory.hpp>
+#include <hpx/traits/polymorphic_traits.hpp>
+#include <hpx/util/static.hpp>
+#include <hpx/util/safe_lexical_cast.hpp>
+#include <hpx/exception.hpp>
+
+#include <boost/preprocessor/stringize.hpp>
+#include <boost/unordered_map.hpp>
+#include <boost/atomic.hpp>
+#include <boost/mpl/bool.hpp>
+
+#include <map>
+
+#include <hpx/config/warnings_prefix.hpp>
+
+namespace hpx { namespace serialization { namespace detail
+{
+    struct type_name_desc_type
+    {
+        BOOST_STATIC_CONSTEXPR boost::uint32_t invalid = ~0u;
+
+        type_name_desc_type(std::string name,
+                boost::uint32_t desc = invalid)
+            : name(std::move(name)), desc(desc)
+        {}
+
+        std::string name;
+        boost::uint32_t desc;
+    };
+
+    BOOST_FORCEINLINE
+    static bool operator<(
+            const hpx::serialization::detail::type_name_desc_type& first,
+            const hpx::serialization::detail::type_name_desc_type& second)
+    {
+        return first.desc < second.desc;
+    }
+
+    class HPX_EXPORT polymorphic_centralized_factory: boost::noncopyable
+    {
+    public:
+        typedef void* (*ctor_type) ();
+        typedef std::multimap<type_name_desc_type, ctor_type> ctor_map_type;
+
+        static void add_factory_function(const std::string& name, ctor_type fun)
+        {
+            if(name.empty())
+            {
+                HPX_THROW_EXCEPTION(serialization_error
+                  , "polymorphic_intrusive_factory::register_class"
+                  , "Cannot register a factory with an empty name");
+            }
+            auto& map = polymorphic_centralized_factory::instance().map_;
+            map.emplace(type_name_desc_type(name), fun);
+        }
+
+        template <class T>
+        static T* create(boost::uint32_t desc)
+        {
+            const auto& map = polymorphic_centralized_factory::instance().map_;
+            auto it = map.find(type_name_desc_type("", desc));
+
+            if (it != map.end())
+                return static_cast<T*>(it->second());
+
+            HPX_THROW_EXCEPTION(serialization_error
+              , "polymorphic_intrusive_factory::create"
+              , "Unknown type descriptor " + util::safe_lexical_cast<std::string>(desc));
+        }
+
+        static std::vector<std::string> get_registered_typenames()
+        {
+            typedef typename ctor_map_type::value_type value_type;
+            const auto& map = polymorphic_centralized_factory::instance().map_;
+
+            std::vector<std::string> result;
+            result.reserve(map.size());
+
+            std::for_each(map.begin(), map.end(),
+                    [&](const value_type& s){
+                        result.push_back(s.first.name);
+                    });
+
+            return result;
```

Comment:
We should not be needing to return copies of all those `string`s that were registered. Once a `string` is registered it would never go away, so as long as we don't move them around we can return pointers, views, string_refs, or whatever feels more appropriate here.